metadata_form/src/js/HierarchicalKeywordSelector.js

```diff
@@ -0,0 +1,237 @@
+import React from 'react'
+import axios from 'axios'
+import { TreeSelect, ConfigProvider, theme } from 'antd'
+
+class HierarchicalKeywordSelector extends React.Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      ...props.formData,
+      // Keyword array in the expected format
+      value: [],
+      treeData: null
+    }
+
+    // Check what theme is set
+    this.colorMode = props.formContext.colorMode
+    this.colorPrimary = getComputedStyle(document.documentElement).getPropertyValue('--bs-primary')
+    this.lightModeSelectColor = getComputedStyle(document.documentElement).getPropertyValue('--bs-black')
+    this.lightModeSelectBgColor = getComputedStyle(document.documentElement).getPropertyValue('--bs-secondary-bg')
+    // From react-select theming
+    this.darkModeSelectColor = '#dee2e6'
+    this.darkModeSelectBgColor = props.formContext.darkThemeColors.primary25
+  }
+
+  async loadData () {
+    const url = this.props.uiSchema['ui:data']
+    const res = await axios(url)
+    return await res.data
+  }
+
+  componentDidMount () {
+    let newVal = []
+    if (this.props.formData) {
+      newVal = this.props.formData.map((keyObj) => {
+        if (Object.keys(keyObj).includes('valueURI')) {
+          return { label: keyObj.Subject, value: [keyObj.Subject, keyObj.valueURI].join(':') }
+        } else {
+          return { label: keyObj.Subject, value: keyObj.Subject + ':' }
+        }
+      })
+    }
+
+    if (!this.state.treeData) {
+      this.loadData().then(data => {
+        const initialTreeData = this.createTreeSelectFromJson(data)
+        const userKeywords = this.getUserKeywords(newVal)
+        const userKeywordTree = this.createUserKeywordTree(userKeywords)
+
+        this.setState({
+          treeData: initialTreeData.concat(userKeywordTree),
+          value: newVal
+        })
+      })
+    }
+  }
+
+  getUserKeywords = (keywords) => {
+    return keywords.filter((keyword) => (keyword.value.endsWith(':')))
+  }
+
+  createUserKeywordTree = (keywords) => {
+    // Given user keywords, create a User keyword tree
+    const newUserCreatedTree = [{
+      title: 'User created keywords',
+      value: 'userCreated'
+      // If disabled is true, pressing enter does not select the keyword.
+      // "disabled": true,
+    }]
+
+    if (keywords.length) {
+      newUserCreatedTree[0].children = keywords.map((keyword) => ({ title: keyword.label, value: keyword.value }))
+      return newUserCreatedTree
+    } else {
+      return []
+    }
+  }
+
+  createTreeSelectFromJson = (jsonA) => {
+    const newTreeJson = []
+    for (const index in jsonA) {
+      const tree = jsonA[index]
+      const newTree = {}
+      newTree.value = tree.text + ':' + tree.extra.uri

```

Comment:
Here I use ":" as a divider between the name of the keyword and the uri, is this acceptable?